bpf/protocol_http2.h

```diff
@@ -95,110 +125,219 @@ http2_grpc_end(http2_conn_stream_t *stream, http2_grpc_request_t *prev_info, voi
         }
 
-        if (pos < (bytes_len - (frame.length + FRAME_HEADER_LEN))) {
-            pos += (frame.length + FRAME_HEADER_LEN);
-            //bpf_dbg_printk("New buf read pos = %d", pos);
+        if (g_ctx->pos < (g_ctx->args.bytes_len - (frame.length + FRAME_HEADER_LEN))) {
+            g_ctx->pos += (frame.length + FRAME_HEADER_LEN);
+            //bpf_dbg_printk("New buf read g_ctx.pos = %d", g_ctx.pos);
         }
+
+        g_ctx->iterations++;
     }
 
-    if (found_start_frame) {
-        http2_grpc_start(
-            &stream, (void *)((u8 *)u_buf + pos), bytes_len, direction, ssl, orig_dport);
-    } else {
-        // We only loop 6 times looking for the stream termination. If the data packed is large we'll miss the
-        // frame saying the stream closed. In that case we try this backup path.
-        if (!found_end_frame && prev_info && saved_stream_id) {
-            if (found_data_frame ||
-                ((prev_info->type == EVENT_HTTP_REQUEST) && (direction == TCP_SEND)) ||
-                ((prev_info->type == EVENT_HTTP_CLIENT) && (direction == TCP_RECV))) {
-                stream.pid_conn = *pid_conn;
-                stream.stream_id = saved_stream_id;
-                found_end_frame = 1;
-            }
-        }
-        if (found_end_frame) {
-            u8 req_type = request_type_by_direction(direction, PACKET_TYPE_RESPONSE);
-            if (prev_info) {
-                if (req_type == prev_info->type) {
-                    u32 buf_pos = saved_buf_pos;
-                    bpf_clamp_umax(buf_pos, IO_VEC_MAX_LEN);
-                    http2_grpc_end(&stream, prev_info, (void *)((u8 *)u_buf + buf_pos));
-                    bpf_map_delete_elem(&active_ssl_connections, pid_conn);
-                } else {
-                    bpf_dbg_printk(
-                        "grpc request/response mismatch, req_type %d, prev_info->type %d",
-                        req_type,
-                        prev_info->type);
-                    bpf_map_delete_elem(&ongoing_http2_grpc, &stream);
-                }
-            }
-        }
+    // this is a weird recursion - we can't loop many times above because the
+    // verifier will reject this program as too complex, we don't want to use
+    // bpf_loop() as we need to support kernels < 5.17, and finally we don't
+    // want to abuse bpf_tail_call as things can get slow (and limited), so we
+    // use this mirror-cracking hybrid approach
+    if (!g_ctx->terminate_search && g_ctx->iterations < 12) {
+        bpf_tail_call(ctx, &jump_table, TAIL_PROTOCOL_HTTP2_GRPC_FRAMES);
+        return 0; // unreachable, but bail safely if bpf_tail_call fails
     }
+
+    // We only loop N times looking for the stream termination. If the data
+    // packed is large we'll miss the frame saying the stream closed. In that
+    // case we try this backup path, which will tail call on success.
+    handle_data_frame(ctx, g_ctx);
+
+    // we found nothing, tough luck!
```

Comment:
Heh, I like the comment but it's also not unusual to find nothing :). HTTP2/gRPC will exchange many frames that are related to the data being sent around, so it's very normal not to find anything except when the transaction starts and ends.

Reply:
Sheer luck then! :) I will nuke that comment.